File: collect_app/src/test/java/org/odk/collect/android/widgets/BarcodeWidgetTest.java
package org.odk.collect.android.widgets;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.equalTo;
import static org.hamcrest.Matchers.nullValue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static org.odk.collect.android.widgets.support.QuestionWidgetHelpers.mockValueChangedListener;
import static org.odk.collect.android.widgets.support.QuestionWidgetHelpers.promptWithAnswer;
import static org.odk.collect.android.widgets.support.QuestionWidgetHelpers.promptWithAppearance;
import static org.odk.collect.android.widgets.support.QuestionWidgetHelpers.promptWithReadOnly;
import static org.odk.collect.android.widgets.support.QuestionWidgetHelpers.widgetTestActivity;
import static org.robolectric.Shadows.shadowOf;

import android.view.View;

import androidx.test.ext.junit.runners.AndroidJUnit4;

import org.javarosa.core.model.FormIndex;
import org.javarosa.core.model.data.StringData;
import org.javarosa.form.api.FormEntryPrompt;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.odk.collect.android.fakes.FakePermissionsProvider;
import org.odk.collect.android.formentry.questions.QuestionDetails;
import org.odk.collect.android.listeners.WidgetValueChangedListener;
import org.odk.collect.android.support.MockFormEntryPromptBuilder;
import org.odk.collect.android.support.WidgetTestActivity;
import org.odk.collect.android.utilities.Appearances;
import org.odk.collect.android.widgets.support.FakeWaitingForDataRegistry;
import org.odk.collect.androidshared.system.CameraUtils;
import org.robolectric.shadows.ShadowActivity;
import org.robolectric.shadows.ShadowToast;

/**
 * @author James Knight
 */
@RunWith(AndroidJUnit4.class)
public class BarcodeWidgetTest {
    private final FakeWaitingForDataRegistry waitingForDataRegistry = new FakeWaitingForDataRegistry();
    private final FakePermissionsProvider permissionsProvider = new FakePermissionsProvider();

    private WidgetTestActivity widgetTestActivity;
    private ShadowActivity shadowActivity;
    private CameraUtils cameraUtils;
    private View.OnLongClickListener listener;
    private FormIndex formIndex;

    @Before
    public void setUp() {
        widgetTestActivity = widgetTestActivity();
        shadowActivity = shadowOf(widgetTestActivity);

        cameraUtils = mock(CameraUtils.class);
        listener = mock(View.OnLongClickListener.class);
        formIndex = mock(FormIndex.class);
        permissionsProvider.setPermissionGranted(true);
    }

    @Test
    public void usingReaDOnly_shouldHideBarcodeButton() {
        assertThat(createWidget(promptWithReadOnly()).binding.barcodeButton.getVisibility(), equalTo(View.GONE));
    }

    @Test
    public void whenPromptHasAnswer_replaceBarcodeButtonIsDisplayed() {
        BarcodeWidget widget = createWidget(promptWithAnswer(new StringData("blah")));
        assertThat(widget.binding.barcodeButton.getText().toString(), equalTo(widgetTestActivity.getString(org.odk.collect.strings.R.string.replace_barcode)));
    }

    @Test
    public void whenPromptHasAnswer_answerTextViewShowsCorrectAnswer() {
        BarcodeWidget widget = createWidget(promptWithAnswer(new StringData("blah")));
        assertThat(widget.binding.barcodeAnswerText.getText().toString(), equalTo("blah"));
    }

    @Test
    public void getAnswer_whenPromptDoesNotHaveAnswer_returnsNull() {
        BarcodeWidget widget = createWidget(promptWithAnswer(null));
        assertThat(widget.getAnswer(), nullValue());
    }

    @Test
    public void getAnswer_whenPromptHasAnswer_returnsCorrectAnswer() {
        BarcodeWidget widget = createWidget(promptWithAnswer(new StringData("blah")));
        assertThat(widget.getAnswer().getDisplayText(), equalTo("blah"));
    }

    @Test
    public void clearAnswer_clearsWidgetAnswer() {
        BarcodeWidget widget = createWidget(promptWithAnswer(new StringData("blah")));
        widget.clearAnswer();

        assertThat(widget.binding.barcodeAnswerText.getText().toString(), equalTo(""));
        assertThat(widget.binding.barcodeButton.getText().toString(), equalTo(widgetTestActivity.getString(org.odk.collect.strings.R.string.get_barcode)));
    }

    @Test
    public void clearAnswer_callsValueChangeListener() {
        BarcodeWidget widget = createWidget(promptWithAnswer(new StringData("blah")));
        WidgetValueChangedListener valueChangedListener = mockValueChangedListener(widget);
        widget.clearAnswer();

        verify(valueChangedListener).widgetValueChanged(widget);
    }

    @Test
    public void setData_updatesWidgetAnswer_afterStrippingInvalidCharacters() {
        BarcodeWidget widget = createWidget(promptWithAnswer(null));
        widget.setData("\ud800blah\b");
        assertThat(widget.binding.barcodeAnswerText.getText().toString(), equalTo("blah"));
    }

    @Test
    public void setData_updatesButtonLabel() {
        BarcodeWidget widget = createWidget(promptWithAnswer(null));
        widget.setData("\ud800blah\b");
        assertThat(widget.binding.barcodeButton.getText(), equalTo(widgetTestActivity.getString(org.odk.collect.strings.R.string.replace_barcode)));
    }

    @Test
    public void setData_callsValueChangeListener() {
        BarcodeWidget widget = createWidget(promptWithAnswer(null));
        WidgetValueChangedListener valueChangedListener = mockValueChangedListener(widget);
        widget.setData("blah");

        verify(valueChangedListener).widgetValueChanged(widget);
    }

    @Test
    public void clickingButtonAndAnswerTextViewForLong_callsLongClickListener() {
        BarcodeWidget widget = createWidget(promptWithAnswer(null));
        widget.setOnLongClickListener(listener);
        widget.binding.barcodeButton.performLongClick();
        widget.binding.barcodeAnswerText.performLongClick();

        verify(listener).onLongClick(widget.binding.barcodeButton);
        verify(listener).onLongClick(widget.binding.barcodeAnswerText);
    }

    @Test
    public void clickingBarcodeButton_whenPermissionIsNotGranted_doesNotLaunchAnyIntent() {
        BarcodeWidget widget = createWidget(promptWithAnswer(null));
        permissionsProvider.setPermissionGranted(false);
        widget.setPermissionsProvider(permissionsProvider);
        widget.binding.barcodeButton.performClick();

        assertThat(shadowActivity.getNextStartedActivity(), nullValue());
        assertThat(waitingForDataRegistry.waiting.isEmpty(), equalTo(true));
    }

    @Test
    public void clickingBarcodeButton_whenPermissionIsGranted_setsWidgetWaitingForData() {
        FormEntryPrompt prompt = promptWithAnswer(null);
        when(prompt.getIndex()).thenReturn(formIndex);

        BarcodeWidget widget = createWidget(prompt);
        widget.setPermissionsProvider(permissionsProvider);
        widget.binding.barcodeButton.performClick();

        assertThat(waitingForDataRegistry.waiting.contains(formIndex), equalTo(true));
    }

    @Test
    public void clickingBarcodeButton_whenFrontCameraIsNotAvailable_showsFrontCameraNotAvailableToast() {
        when(cameraUtils.isFrontCameraAvailable(any())).thenReturn(false);
        BarcodeWidget widget = createWidget(promptWithAppearance(Appearances.FRONT));
        widget.setPermissionsProvider(permissionsProvider);
        widget.binding.barcodeButton.performClick();

        assertThat(ShadowToast.getTextOfLatestToast(), equalTo(widgetTestActivity.getString(org.odk.collect.strings.R.string.error_front_camera_unavailable)));
    }

    @Test
    public void clickingBarcodeButton_whenFrontCameraIsAvailable_launchesCorrectIntent() {
        when(cameraUtils.isFrontCameraAvailable(any())).thenReturn(true);
        BarcodeWidget widget = createWidget(promptWithAppearance(Appearances.FRONT));
        widget.setPermissionsProvider(permissionsProvider);
        widget.binding.barcodeButton.performClick();

        assertThat(shadowActivity.getNextStartedActivity().getBooleanExtra(Appearances.FRONT, false), equalTo(true));
    }

    @Test
    public void whenPromptHasHiddenAnswerAppearance_answerIsNotDisplayed() {
        FormEntryPrompt prompt = new MockFormEntryPromptBuilder(promptWithAppearance(Appearances.HIDDEN_ANSWER))
                .withAnswer(new StringData("original contents"))
                .build();

        BarcodeWidget widget = createWidget(prompt);

        // Check initial value is not shown
        assertThat(widget.binding.barcodeAnswerText.getVisibility(), equalTo(View.GONE));
        assertThat(widget.binding.barcodeButton.getText(), equalTo(widgetTestActivity.getString(org.odk.collect.strings.R.string.replace_barcode)));
        assertThat(widget.getAnswer(), equalTo(new StringData("original contents")));

        // Check updates aren't shown
        widget.setData("updated contents");
        assertThat(widget.binding.barcodeAnswerText.getVisibility(), equalTo(View.GONE));
        assertThat(widget.binding.barcodeButton.getText(), equalTo(widgetTestActivity.getString(org.odk.collect.strings.R.string.replace_barcode)));
        assertThat(widget.getAnswer(), equalTo(new StringData("updated contents")));
    }

    public BarcodeWidget createWidget(FormEntryPrompt prompt) {
        return new BarcodeWidget(widgetTestActivity, new QuestionDetails(prompt),
                waitingForDataRegistry, cameraUtils);
    }
}

File: collect_app/src/test/java/org/odk/collect/android/widgets/BarcodeWidgetTest.kt
package org.odk.collect.android.widgets

import androidx.compose.ui.test.junit4.createAndroidComposeRule
import androidx.compose.ui.test.onNodeWithText
import androidx.compose.ui.test.performClick
import net.bytebuddy.utility.RandomString
import org.hamcrest.CoreMatchers.equalTo
import org.hamcrest.MatcherAssert.assertThat
import org.javarosa.core.model.Constants
import org.javarosa.core.model.data.StringData
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.mockito.ArgumentMatchers
import org.mockito.Mockito.verify
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever
import org.odk.collect.android.fakes.FakePermissionsProvider
import org.odk.collect.android.formentry.questions.QuestionDetails
import org.odk.collect.android.support.MockFormEntryPromptBuilder
import org.odk.collect.android.support.WidgetTestActivity
import org.odk.collect.android.utilities.Appearances
import org.odk.collect.android.widgets.barcode.BarcodeWidget
import org.odk.collect.android.widgets.base.QuestionWidgetTest
import org.odk.collect.android.widgets.support.FakeWaitingForDataRegistry
import org.odk.collect.android.widgets.support.QuestionWidgetHelpers
import org.odk.collect.androidshared.system.CameraUtils
import org.odk.collect.androidtest.onNodeWithClickLabel
import org.odk.collect.strings.R.string
import org.robolectric.Shadows
import org.robolectric.shadows.ShadowToast

class BarcodeWidgetTest : QuestionWidgetTest<BarcodeWidget, StringData>() {
    @get:Rule
    val composeRule = createAndroidComposeRule<WidgetTestActivity>()
    private val waitingForDataRegistry = FakeWaitingForDataRegistry()
    private val permissionsProvider = FakePermissionsProvider().apply {
        setPermissionGranted(true)
    }
    private val cameraUtils = mock<CameraUtils>()

    override fun createWidget(): BarcodeWidget {
        return BarcodeWidget(
            composeRule.activity,
            QuestionDetails(formEntryPrompt),
            dependencies,
            waitingForDataRegistry,
            cameraUtils
        ).also {
            composeRule.activity.setContentView(it)
            activity = composeRule.activity
        }
    }

    override fun getNextAnswer(): StringData? {
        return StringData(RandomString.make())
    }

    @Before
    fun setup() {
        formEntryPrompt = MockFormEntryPromptBuilder()
            .withControlType(Constants.CONTROL_INPUT)
            .withDataType(Constants.DATATYPE_BARCODE)
            .build()
    }

    @Test
    fun `The button is hidden in read-only mode`() {
        formEntryPrompt = MockFormEntryPromptBuilder(formEntryPrompt)
            .withReadOnly(true)
            .build()
        createWidget()
        composeRule.onNodeWithClickLabel(activity.getString(string.get_barcode)).assertDoesNotExist()
    }

    @Test
    fun `Display the 'Replace Barcode' button if answer is present`() {
        formEntryPrompt = MockFormEntryPromptBuilder(formEntryPrompt)
            .withAnswer(StringData("blah"))
            .build()
        createWidget()
        composeRule.onNodeWithClickLabel(activity.getString(string.replace_barcode)).assertExists()
    }

    @Test
    fun `Display the answer if answer is present`() {
        formEntryPrompt = MockFormEntryPromptBuilder(formEntryPrompt)
            .withAnswer(StringData("blah"))
            .build()
        createWidget()
        composeRule.onNodeWithText("blah").assertExists()
    }

    @Test
    fun `#getAnswer returns null when there is no answer`() {
        val widget = createWidget()
        assertThat(widget.answer, equalTo(null))
    }

    @Test
    fun `#getAnswer returns the answer when there is answer`() {
        formEntryPrompt = MockFormEntryPromptBuilder(formEntryPrompt)
            .withAnswer(StringData("blah"))
            .build()
        val widget = createWidget()
        assertThat(widget.answer!!.displayText, equalTo("blah"))
    }

    @Test
    fun `#clearAnswer removes answer and updates button title`() {
        formEntryPrompt = MockFormEntryPromptBuilder(formEntryPrompt)
            .withAnswer(StringData("blah"))
            .build()
        val widget = createWidget()
        widget.clearAnswer()

        assertThat(
            widget.answer,
            equalTo(null)
        )
        composeRule.onNodeWithClickLabel(activity.getString(string.get_barcode)).assertExists()
    }

    @Test
    fun `#clearAnswer calls #valueChangeListener`() {
        formEntryPrompt = MockFormEntryPromptBuilder(formEntryPrompt)
            .withAnswer(StringData("blah"))
            .build()
        val widget = createWidget()
        val valueChangedListener = QuestionWidgetHelpers.mockValueChangedListener(widget)
        widget.clearAnswer()

        verify(valueChangedListener).widgetValueChanged(widget)
    }

    @Test
    fun `#setData displays sanitized answer`() {
        val widget = createWidget()
        widget.setData("\ud800blah\b")
        assertThat(
            widget.answer!!.displayText,
            equalTo("blah")
        )
    }

    @Test
    fun `#setData updates the button title`() {
        val widget = createWidget()
        widget.setData("\ud800blah\b")
        composeRule.onNodeWithClickLabel(activity.getString(string.replace_barcode)).assertExists()
    }

    @Test
    fun `#setData calls #valueChangeListener`() {
        val widget = createWidget()
        val valueChangedListener = QuestionWidgetHelpers.mockValueChangedListener(widget)
        widget.setData("blah")

        verify(valueChangedListener).widgetValueChanged(widget)
    }

    @Test
    fun `pressing the button with permission not granted does not launch anything`() {
        val widget = createWidget()
        permissionsProvider.setPermissionGranted(false)
        widget.setPermissionsProvider(permissionsProvider)
        composeRule.onNodeWithClickLabel(activity.getString(string.get_barcode)).performClick()

        assertThat(Shadows.shadowOf(activity).nextStartedActivity, equalTo(null))
        assertThat(waitingForDataRegistry.waiting.isEmpty(), equalTo(true))
    }

    @Test
    fun `pressing the button with permission granted registers widget for data waiting`() {
        whenever(formEntryPrompt.index).thenReturn(formIndex)

        val widget = createWidget()
        widget.setPermissionsProvider(permissionsProvider)
        composeRule.onNodeWithClickLabel(activity.getString(string.get_barcode)).performClick()

        assertThat(
            waitingForDataRegistry.waiting.contains(formIndex),
            equalTo(true)
        )
    }

    @Test
    fun `pressing the button when front camera should be used but it is not available displays a toast`() {
        whenever(cameraUtils.isFrontCameraAvailable(ArgumentMatchers.any())).thenReturn(false)
        formEntryPrompt = MockFormEntryPromptBuilder(formEntryPrompt)
            .withAppearance(Appearances.FRONT)
            .build()
        val widget = createWidget()
        widget.setPermissionsProvider(permissionsProvider)
        composeRule.onNodeWithClickLabel(activity.getString(string.get_barcode)).performClick()

        assertThat(
            ShadowToast.getTextOfLatestToast(),
            equalTo(activity.getString(string.error_front_camera_unavailable))
        )
    }

    @Test
    fun `pressing the button when front camera should be used and it is available launches correct intent`() {
        whenever(cameraUtils.isFrontCameraAvailable(ArgumentMatchers.any())).thenReturn(true)
        formEntryPrompt = MockFormEntryPromptBuilder(formEntryPrompt)
            .withAppearance(Appearances.FRONT)
            .build()
        val widget = createWidget()
        widget.setPermissionsProvider(permissionsProvider)
        composeRule.onNodeWithClickLabel(activity.getString(string.get_barcode)).performClick()

        assertThat(
            Shadows.shadowOf(activity).nextStartedActivity.getBooleanExtra(Appearances.FRONT, false),
            equalTo(true)
        )
    }

    @Test
    fun `The answer is not displayed with hidden mode`() {
        formEntryPrompt = MockFormEntryPromptBuilder(formEntryPrompt)
            .withAppearance(Appearances.HIDDEN_ANSWER)
            .withAnswer(StringData("original contents"))
            .build()

        val widget = createWidget()

        // Check initial value is not shown
        composeRule.onNodeWithClickLabel(activity.getString(string.replace_barcode)).assertExists()
        composeRule.onNodeWithText("original contents").assertDoesNotExist()
        assertThat(widget.answer, equalTo(StringData("original contents")))

        // Check updates aren't shown
        widget.setData("updated contents")
        composeRule.onNodeWithClickLabel(activity.getString(string.replace_barcode)).assertExists()
        composeRule.onNodeWithText("updated contents").assertDoesNotExist()
        assertThat(
            widget.answer,
            equalTo(StringData("updated contents"))
        )
    }

    override fun usingReadOnlyOptionShouldMakeAllClickableElementsDisabled() {
        formEntryPrompt = MockFormEntryPromptBuilder(formEntryPrompt)
            .withReadOnly(true)
            .build()
        createWidget()
        composeRule.onNodeWithClickLabel(activity.getString(string.get_barcode)).assertDoesNotExist()
    }
}
